<a href="https://colab.research.google.com/github/PedroH-07/Checkpoint2-SERS/blob/main/Checkpoint2_SERS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Appliance Energy Prediction**

In [15]:
#Importando Bibliotecas
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import numpy as np

In [ ]:
# colocando o dataset e pritando as informações dele
df = pd.read_csv('/content/energydata_complete.csv')

print(df.head())
print(df.info())
print(df.describe())

                  date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility

In [ ]:
X = df.drop(['date', 'Appliances'], axis=1)
y = df['Appliances']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Regressão Linear
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)

In [ ]:
#Árvore de Regressão (Decision Tree)
dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)

In [ ]:
#Random Forest
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

In [ ]:
# Avaliação dos modelos usando R², RMSE e MAE
def print_metrics(y_true, y_pred, model_name):
    r2 = r2_score(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mae = mean_absolute_error(y_true, y_pred)
    print(f"\n{model_name}:")
    print(f"R² = {r2:.4f}")
    print(f"RMSE = {rmse:.2f}")
    print(f"MAE = {mae:.2f}")

In [ ]:
print_metrics(y_test, y_pred_lr, "Regressão Linear")
print_metrics(y_test, y_pred_dt, "Árvore de Decisão")
print_metrics(y_test, y_pred_rf, "Random Forest")


Regressão Linear:
R² = 0.1693
RMSE = 91.17
MAE = 52.55

Árvore de Decisão:
R² = 0.1798
RMSE = 90.60
MAE = 38.87

Random Forest:
R² = 0.5322
RMSE = 68.42
MAE = 32.90


Conclusão:

- **Random Forest** tem o maior valor de R² (0.5322), indicando que explica uma proporção maior da variância no consumo de energia em comparação com a Regressão Linear (0.1693) e a Árvore de Decisão (0.1798).
- **Random Forest** também tem os menores valores de RMSE (68.42) e MAE (32.90), o que significa que suas previsões estão, em média, mais próximas dos valores reais do que os outros dois modelos.

Portanto, o modelo **Random Forest** é o melhor entre os três para explicar e prever o consumo de energia com base nesta análise.

# **Smart Grid Stability**

In [ ]:
df_stability = pd.read_csv('smart_grid_stability_augmented.csv')
print(df_stability.head())
print(df_stability.info())

       tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Data columns (total 14 column

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

# Decision Tree Classifier
dt_stability = DecisionTreeClassifier(random_state=42)
dt_stability.fit(X_train_stability, y_train_stability)
y_pred_dt_stability = dt_stability.predict(X_test_stability)

# KNN Classifier
knn_stability = KNeighborsClassifier()
knn_stability.fit(X_train_stability, y_train_stability)
y_pred_knn_stability = knn_stability.predict(X_test_stability)

# Logistic Regression Classifier
lr_stability = LogisticRegression(random_state=42, max_iter=1000)
lr_stability.fit(X_train_stability, y_train_stability)
y_pred_lr_stability = lr_stability.predict(X_test_stability)

def print_classification_metrics(y_true, y_pred, model_name):
    accuracy = accuracy_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred, pos_label='stable')
    cm = confusion_matrix(y_true, y_pred)

    print(f"\n{model_name}:")
    print(f"Accuracy = {accuracy:.4f}")
    print(f"F1-score = {f1:.4f}")
    print("Confusion Matrix:")
    print(cm)

## Comparação de Modelos e Detecção de Instabilidade

Com base nas métricas de avaliação obtidas (Acurácia, F1-score e Matriz de Confusão), podemos comparar o desempenho dos modelos de classificação na detecção da estabilidade da rede elétrica:

- **Random Forest:** Apresentou o melhor desempenho geral, com a maior acurácia e F1-score. Sua matriz de confusão mostra um número menor de falsos positivos e falsos negativos em comparação com os outros modelos, indicando que ele é mais eficaz em classificar corretamente tanto instâncias estáveis quanto instáveis. Isso o torna o modelo mais confiável para detectar instabilidade neste conjunto de dados.

- **Árvore de Decisão:** Teve um desempenho razoável, mas significativamente inferior ao Random Forest em todas as métricas.

- **KNN e Regressão Logística:** Apresentaram os menores valores de acurácia e F1-score, e suas matrizes de confusão indicam um número maior de erros de classificação.

**Conclusão:**

Considerando as métricas de desempenho, especialmente o F1-score (que é uma boa medida para problemas de classificação com classes desbalanceadas, se for o caso) e a análise da matriz de confusão, o modelo **Random Forest** é o mais confiável para detectar instabilidade na rede elétrica com base neste conjunto de dados. Sua capacidade de generalização e menor número de erros de classificação o destacam como a melhor opção entre os modelos testados.

In [ ]:
print_classification_metrics(y_test_stability, y_pred_dt_stability, "Árvore de Decisão")
print_classification_metrics(y_test_stability, y_pred_knn_stability, "KNN")
print_classification_metrics(y_test_stability, y_pred_lr_stability, "Regressão Logística")
print_classification_metrics(y_test_stability, y_pred_stability, "Random Forest")


Árvore de Decisão:
Accuracy = 0.8958
F1-score = 0.8556
Confusion Matrix:
[[3705  617]
 [ 634 7044]]

KNN:
Accuracy = 0.8247
F1-score = 0.7454
Confusion Matrix:
[[3079 1243]
 [ 860 6818]]

Regressão Logística:
Accuracy = 0.8101
F1-score = 0.7249
Confusion Matrix:
[[3003 1319]
 [ 960 6718]]

Random Forest:
Accuracy = 0.9481
F1-score = 0.9270
Confusion Matrix:
[[3955  367]
 [ 256 7422]]


In [ ]:
X_stability = df_stability.drop(['stab', 'stabf'], axis=1)
y_stability = df_stability['stabf']
X_train_stability, X_test_stability, y_train_stability, y_test_stability = train_test_split(X_stability, y_stability, test_size=0.2, random_state=42)
display(X_train_stability.head())
display(y_train_stability.head())

tau1      tau2      tau3      tau4        p1        p2        p3  \
48572  8.347672  9.984817  8.155656  7.135798  4.695577 -1.896789 -1.700429   
38696  9.895374  5.218368  1.049134  1.889521  2.854356 -0.687905 -0.544313   
13611  7.805887  1.425355  3.627628  6.967638  3.030196 -0.779081 -1.638151   
35213  7.757639  0.944923  9.625876  7.343146  4.358784 -1.889955 -0.779443   
31766  0.721472  2.519541  5.495173  7.852192  5.169904 -1.556343 -1.667228   

             p4        g1        g2        g3        g4  
48572 -1.098359  0.944713  0.968412  0.457960  0.757765  
38696 -1.622138  0.401010  0.406740  0.336036  0.662520  
13611 -0.612964  0.360131  0.829569  0.282305  0.079182  
35213 -1.689386  0.284492  0.176676  0.296432  0.184116  
31766 -1.946333  0.358372  0.328424  0.452532  0.980840

48572    unstable
38696      stable
13611      stable
35213      stable
31766    unstable
Name: stabf, dtype: object

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_stability = RandomForestClassifier(random_state=42)
rf_stability.fit(X_train_stability, y_train_stability)

RandomForestClassifier(random_state=42)

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred_stability = rf_stability.predict(X_test_stability)

accuracy = accuracy_score(y_test_stability, y_pred_stability)
precision = precision_score(y_test_stability, y_pred_stability, pos_label='stable')
recall = recall_score(y_test_stability, y_pred_stability, pos_label='stable')
f1 = f1_score(y_test_stability, y_pred_stability, pos_label='stable')

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.9481
Precision: 0.9392
Recall: 0.9151
F1-score: 0.9270


In [20]:
df = pd.read_csv('SolarPrediction.csv')

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix
import pandas as pd
import numpy as np

mediana_radiacao = df['Radiation'].median()
df['Radiação_Alta'] = (df['Radiation'] > mediana_radiacao).astype(int)

# Removendo as colunas não relevantes
features = ['UNIXTime', 'Temperature', 'Pressure', 'Humidity', 'WindDirection(Degrees)', 'Speed']
X = df[features]
y = df['Radiação_Alta']

# Separar dados em treino e teste (70/30)
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalizar atributos contínuos
scaler = StandardScaler()
X_treino_normalizado = scaler.fit_transform(X_treino)
X_teste_normalizado = scaler.transform(X_teste)

modelos = {
    "ÁRVORE DE DECISÃO": DecisionTreeClassifier(random_state=42),
    "RANDOM FOREST": RandomForestClassifier(random_state=42),
    "SVM": SVC(random_state=42)
}

resultados = {}

for nome_modelo, modelo in modelos.items():
    # Treinar o modelo
    modelo.fit(X_treino_normalizado, y_treino)

    # Fazer previsões
    y_predito = modelo.predict(X_teste_normalizado)

    # Avaliar o modelo
    acuracia = accuracy_score(y_teste, y_predito)
    matriz_confusao = confusion_matrix(y_teste, y_predito)

    resultados[nome_modelo] = {
        "Acurácia": acuracia,
        "Matriz de Confusão": matriz_confusao
    }

print("Resultados da Classificação de Radiação Solar:\n")

for nome_modelo, metricas in resultados.items():
    print(f"--- {nome_modelo} ---")
    print(f"Acurácia: {metricas['Acurácia']:.4f}")
    print("Matriz de Confusão:")
    print(metricas['Matriz de Confusão'])
    print("-" * (len(nome_modelo) + 6))

Resultados da Classificação de Radiação Solar:

--- ÁRVORE DE DECISÃO ---
Acurácia: 0.9436
Matriz de Confusão:
[[4669  284]
 [ 269 4584]]
-----------------------
--- RANDOM FOREST ---
Acurácia: 0.9491
Matriz de Confusão:
[[4783  170]
 [ 329 4524]]
-------------------
--- SVM ---
Acurácia: 0.8452
Matriz de Confusão:
[[4680  273]
 [1245 3608]]
---------
